File: core/infra/src/test/java/com/beat/infra/persistence/performance/mapper/PerformancePersistenceMapperTest.java
package com.beat.infra.persistence.performance.mapper;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

import java.time.LocalDate;

import org.junit.jupiter.api.Test;

import com.beat.domain.sharedkernel.vo.BankName;
import com.beat.domain.performance.model.Genre;
import com.beat.domain.performance.vo.PaymentAccount;
import com.beat.domain.performance.model.Performance;
import com.beat.domain.performance.vo.PerformancePeriod;
import com.beat.domain.performance.vo.RunningTime;
import com.beat.domain.performance.vo.TicketPrice;
import com.beat.infra.persistence.exception.PersistenceMappingException;
import com.beat.infra.persistence.performance.entity.PerformanceJpaEntity;
import com.beat.infra.persistence.performance.entity.PerformancePeriodJpaValue;

class PerformancePersistenceMapperTest {
	private final PerformancePersistenceMapper mapper = new PerformancePersistenceMapper();

	@Test
	void roundTripsValueObjectsWithoutChangingLegacyColumns() {
		Performance domain = performance(PaymentAccount.of(BankName.KAKAOBANK, "123", "holder"));

		PerformanceJpaEntity entity = mapper.toEntity(domain);
		Performance roundTrip = mapper.toDomain(entity);

		assertEquals(BankName.KAKAOBANK, entity.getPaymentAccount().getBankName());
		assertEquals(LocalDate.of(2026, 7, 16), entity.getPerformancePeriodValue().getStartDate());
		assertEquals("2026.07.16~2026.07.18", entity.getLegacyPerformancePeriod());
		assertEquals(domain.getPaymentAccount(), roundTrip.getPaymentAccount());
		assertEquals(domain.getPerformancePeriodValue(), roundTrip.getPerformancePeriodValue());
		assertEquals(domain.getRunningTimeValue(), roundTrip.getRunningTimeValue());
		assertEquals(domain.getTicketPriceValue(), roundTrip.getTicketPriceValue());
	}

	@Test
	void keepsAllNullPaymentAccountAsNull() {
		PerformanceJpaEntity entity = mapper.toEntity(performance(null));

		assertNull(entity.getPaymentAccount());
		assertNull(mapper.toDomain(entity).getPaymentAccount());
	}

	@Test
	void rejectsPartiallyPopulatedPeriodColumns() throws ReflectiveOperationException {
		PerformanceJpaEntity entity = mapper.toEntity(performance(null));
		PerformancePeriodJpaValue partialPeriod = new PerformancePeriodJpaValue(
			LocalDate.of(2026, 7, 16), LocalDate.of(2026, 7, 18));
		var startDateField = PerformancePeriodJpaValue.class.getDeclaredField("startDate");
		startDateField.setAccessible(true);
		startDateField.set(partialPeriod, null);
		entity = PerformanceJpaEntity.rehydrate(
			entity.getId(), entity.getPerformanceTitle(), entity.getGenre(), entity.getRunningTime(),
			entity.getPerformanceDescription(), entity.getPerformanceAttentionNote(), entity.getPaymentAccount(),
			entity.getPosterImage(), entity.getPerformanceTeamName(), entity.getPerformanceVenue(),
			entity.getRoadAddressName(), entity.getPlaceDetailAddress(), entity.getLatitude(), entity.getLongitude(),
			entity.getPerformanceContact(), partialPeriod,
			entity.getLegacyPerformancePeriod(), entity.getTicketPrice(), entity.getTotalScheduleCount(), entity.getUserId());

		PerformanceJpaEntity partiallyPopulated = entity;
		assertThrows(PersistenceMappingException.class, () -> mapper.toDomain(partiallyPopulated));
	}

	private Performance performance(PaymentAccount paymentAccount) {
		return Performance.rehydrate(
			1L, "title", Genre.BAND, RunningTime.of(90), "description", "attention", paymentAccount,
			"poster", "team", "venue", "road", "detail", "37.1", "127.1", "010-0000-0000",
			PerformancePeriod.of(LocalDate.of(2026, 7, 16), LocalDate.of(2026, 7, 18)),
			TicketPrice.of(20_000), 3, 7L
		);
	}
}

File: infrastructure/src/test/kotlin/com/beat/infrastructure/persistence/performance/mapper/PerformancePersistenceMapperTest.kt
package com.beat.infrastructure.persistence.performance.mapper

import com.beat.domain.performance.model.Genre
import com.beat.domain.performance.model.Performance
import com.beat.domain.performance.vo.PaymentAccount
import com.beat.domain.performance.vo.PerformancePeriod
import com.beat.domain.performance.vo.RunningTime
import com.beat.domain.performance.vo.TicketPrice
import com.beat.domain.sharedkernel.vo.BankName
import io.kotest.core.spec.style.FunSpec
import io.kotest.matchers.shouldBe
import java.time.LocalDate

class PerformancePersistenceMapperTest :
    FunSpec({
        val mapper = PerformancePersistenceMapper()

        test("레거시 컬럼을 변경하지 않고 value object를 왕복한다") {
            val domain = performance(PaymentAccount.of(BankName.KAKAOBANK, "123", "holder"))

            val entity = mapper.toEntity(domain)
            val roundTrip = mapper.toDomain(entity)

            entity.paymentAccount!!.bankName shouldBe BankName.KAKAOBANK
            entity.performancePeriodValue.startDate shouldBe LocalDate.of(2026, 7, 16)
            roundTrip.paymentAccount shouldBe domain.paymentAccount
            roundTrip.performancePeriodValue shouldBe domain.performancePeriodValue
            roundTrip.runningTimeValue shouldBe domain.runningTimeValue
            roundTrip.ticketPriceValue shouldBe domain.ticketPriceValue
        }

        test("모두 null인 payment account는 null로 유지된다") {
            val entity = mapper.toEntity(performance(null))

            entity.paymentAccount shouldBe null
            mapper.toDomain(entity).paymentAccount shouldBe null
        }

        test("CONTRACT 이후 period 컬럼은 항상 not null로 저장된다") {
            val entity = mapper.toEntity(performance(null))

            entity.performancePeriodValue.startDate shouldBe LocalDate.of(2026, 7, 16)
            entity.performancePeriodValue.endDate shouldBe LocalDate.of(2026, 7, 18)
        }
    })

private fun performance(paymentAccount: PaymentAccount?): Performance =
    Performance.rehydrate(
        1L,
        "title",
        Genre.BAND,
        RunningTime.of(90),
        "description",
        "attention",
        paymentAccount,
        "poster",
        "team",
        "venue",
        "road",
        "detail",
        "37.1",
        "127.1",
        "010-0000-0000",
        PerformancePeriod.of(LocalDate.of(2026, 7, 16), LocalDate.of(2026, 7, 18)),
        TicketPrice.of(20_000),
        3,
        7L,
    )
